File: Week 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <string.h>

bool isValid(long n, string type);
string getType(long n);
int extract(long n, int place);
int getN(long n);
bool checksum(long n);

int main(void)
{
    string type;
    long n;

    // ask for the number
    n = get_long("Number: ");

    // figures what type of card
    type = getType(n);

    if (strcmp(type, "INVALID") == 0)
    {
        printf("INVALID\n");
    }
    else
    {
        // is it valid?
        if (isValid(n, type))
        {
            printf("%s\n", type);
        }
        else
        {
            printf("INVALID\n");
        }
    }
}

// checks the validity through n length and numcheck
bool isValid(long n, string type)
{
    if (strcmp(type, "VISA") == 0 && (getN(n) == 13 || getN(n) == 16))
    {
        return true;
    }
    else if (strcmp(type, "AMEX") == 0 && getN(n) == 15)
    {
        return true;
    }
    else if (strcmp(type, "MASTERCARD") == 0 && getN(n) == 16)
    {
        return true;
    }
    else
    {
        return false;
    }
}

// gets what type of card
// visa, amex, or mastercard
string getType(long n)
{
    // visa check
    if (extract(n, 10) == 4 && checksum(n))
    {
        return "VISA";
    }
    // amex check
    else if ((extract(n, 100) == 34 || extract(n, 100) == 37) && checksum(n))
    {
        return "AMEX";
    }
    // mastercard check
    else if ((extract(n, 100) >= 51 && extract(n, 100) <= 55) && checksum(n))
    {
        return "MASTERCARD";
    }
    else
    {
        return "INVALID";
    }
}

// gets the specific number
// 10 == first digit, 100 == first 2 digits
int extract(long n, int place)
{
    while (n >= place)
    {
        n = n / 10;
    }
    return (int)n;
}

// gets the n length
int getN(long n)
{
    return log10(n) + 1;
}

// luhn's algorithm
bool checksum(long n)
{
    int sum = 0;
    int num_digits = 0;
    int parity = 0;
    long digits = n;

    while (digits > 0)
    {
        // checks if the current digit is in an odd position or an even position.
        if (parity == 1)
        {
            int digit = digits % 10;
            digit = digit * 2;
            if (digit > 9)
            {
                digit = digit - 9;
            }
            sum += digit;
        }
        else
        {
            sum += digits % 10;
        }
        digits /= 10;
        num_digits++;
        parity = 1 - parity;
    }

    if ((sum % 10) == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

File: Week 6/sentimental-credit/credit.py
# TODO
from cs50 import get_int


def main():
    # ask for the number
    n = get_int("Number: ")

    type = getType(n)
    if (type == "INVALID"):
        print("INVALID")
    else:
        if (isValid(n, type)):
            print(type)
        else:
            print("INVALID")

# check the validity


def isValid(n, type):
    if (type == "VISA" and (len(str(n)) == 13 or len(str(n)) == 16)):
        return True
    elif (type == "AMEX" and len(str(n)) == 15):
        return True
    elif (type == "MASTERCARD" and len(str(n)) == 16):
        return True
    else:
        return False

# get what type of card


def getType(n):
    # visa
    if (int(str(n)[0]) == 4 and checksum(n)):
        return "VISA"
    # amex
    elif ((int(str(n)[:2]) == 34 or int(str(n)[:2]) == 37) and checksum(n)):
        return "AMEX"
    # mastercard
    elif ((int(str(n)[:2]) >= 51 and int(str(n)[:2]) <= 55) and checksum(n)):
        return "MASTERCARD"
    else:
        return "INVALID"

# luhn's algorithm


def checksum(n):
    # multiply every other digit by 2, starting with the number’s second-to-last digit,
    # and then add those products’ digits together.
    sum = 0
    for i in range(len(str(n)) - 2, -1, -2):
        sum += int(str(n)[i]) * 2
        if (int(str(n)[i]) * 2 > 9):
            sum += 1
    # add the sum to the sum of the digits that weren’t multiplied by 2.
    for i in range(len(str(n)) - 1, -1, -2):
        sum += int(str(n)[i])
    # if the total’s last digit is 0(or, put more formally, if the total modulo 10 is congruent to 0),
    # the number is valid!
    if (sum % 10 == 0):
        return True
    else:
        return False


main()
